caikit/core/model_management/local_model_loader.py

```diff
@@ -0,0 +1,243 @@
+            self._configure_backend_overrides(backend_type, backend_instance)
+
+            # Add the instance to the registry
+            self._backends.append(backend_instance)
+
+        log.debug2("All configured backends: %s", self._backends)
+
+    def load(self, model_config: ModuleConfig, **kwargs) -> Optional[ModuleBase]:
+        """Given a ModelConfig, attempt to load it into memory
+
+        Args:
+            model_config (ModuleConfig): The in-memory model config object for
+                the model to be loaded
+
+        Returns:
+            model (Optional[ModuleBase]): The in-memory ModuleBase instance that
+                is ready to run
+        """
+        module_id = model_config.module_id
+        model_path = model_config.model_path
+        module_implementations = module_backend_registry().get(module_id, {})
+        log.debug2(
+            "Number of available backend implementations for %s found: %d",
+            module_id,
+            len(module_implementations),
+        )
+        # Look up the backend that this model was created with
+        model_creation_backend = model_config.get("model_backend", backend_types.LOCAL)
+        log.debug2("Model creation backend: %s", model_creation_backend)
+
+        # Iterate through each backend in priority order and see if this module
+        # can load with it
+        loaded_model = None
+        for load_backend in self._backends:
+
+            # Look in the module's implementations for this backend type
+            backend_impl_obj = module_implementations.get(load_backend.backend_type)
+            if backend_impl_obj is None:
+                log.debug3(
+                    "Module %s does not support loading with %s",
+                    module_id,
+                    load_backend.backend_type,
+                )
+                continue
+
+            # Grab the concrete module class for this backend and check to
+            # see if this model's artifacts were created with a version of
+            # the module that can be loaded with this backend.
+            module_backend_impl = backend_impl_obj.impl_class
+            supported_load_backends = self._get_supported_load_backends(
+                module_backend_impl
+            )
+            if model_creation_backend in supported_load_backends:
+                log.debug3(
+                    "Attempting to load %s (module_id %s) with backend %s and class %s",
+                    model_path,
+                    module_id,
+                    load_backend.backend_type,
+                    module_backend_impl.__name__,
+                )
+                extra_kwargs = {}
+                if self._supports_load_backend_kwarg(module_backend_impl.load):
+                    extra_kwargs["load_backend"] = load_backend
+                loaded_model = module_backend_impl.load(
+                    model_path,
+                    **extra_kwargs,
+                    **kwargs,
+                )
```

Comment:
This was a corner that I had to do some spelunking to sort through which lead me to this PR: https://github.com/caikit/caikit/pull/129

The gist is that often a module needs to know its backed _during_ the `__init__`/`load` of the module itself rather than during `run`, so providing it as a kwarg goes to the `load` function allows the logic of initialization to have a handle to the backend.

The previous solution was _always_ passing this kwarg, but we were just getting lucky that in the case of a model being loaded locally, it was never getting to this line because the `LocalSharedModelLoadBackend` was getting to it first.